testsuite/cluster-test/src/health/fullnode_check.rs

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use crate::{
+    cluster::Cluster,
+    health::{HealthCheck, HealthCheckContext},
+};
+use async_trait::async_trait;
+use once_cell::sync::Lazy;
+use regex::Regex;
+use std::collections::HashMap;
+
+static VAL_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"val-(\d+)").unwrap());
+static FULLNODE_REGEX: Lazy<Regex> = Lazy::new(|| Regex::new(r"fn-(\d+)").unwrap());
+
+pub struct FullNodeHealthCheck {
+    cluster: Cluster,
+}
+
+impl FullNodeHealthCheck {
+    pub fn new(cluster: Cluster) -> FullNodeHealthCheck {
+        Self { cluster }
+    }
+}
+
+pub fn node_index(peer_name: String) -> String {
+    if let Some(cap) = VAL_REGEX.captures(peer_name.as_ref()) {
+        if let Some(cap) = cap.get(1) {
+            return cap.as_str().to_string();
+        }
+    }
+    if let Some(cap) = FULLNODE_REGEX.captures(peer_name.as_ref()) {
+        if let Some(cap) = cap.get(1) {
+            return cap.as_str().to_string();
+        }
+    }
+    panic!(
+        "Failed to parse peer name {} into validator_index",
+        peer_name
+    )
+}
+
+#[async_trait]
+impl HealthCheck for FullNodeHealthCheck {
+    async fn verify(&mut self, ctx: &mut HealthCheckContext) {
+        let validator = self.cluster.validator_instances();
+        let fullnode = self.cluster.fullnode_instances();
+        let mut val_latest_version = HashMap::new();
+        for v in validator {
+            let res = v
+                .debug_interface_client()
+                .get_node_metric("libra_state_sync_committed_version{}")
+                .await;
+            let version = match res {
+                Ok(res) => res.unwrap_or(0i64),
+                _ => 0i64,
+            };
+            if version != 0 {
+                val_latest_version.insert(node_index(v.peer_name().to_string()), version);
+            }
+        }
+        for v in fullnode {
```

Comment:
Similarly to above(without converting to map indeed), you can make query all full nodes in parallel and make list of tuples of `(full node, version)`.

After you got results, you can just iterate over this list to do the check - this step don't need to be parallelized

Goal is to parallelize network calls to `get_node_metric`, everything else can be done sequentially